lib/Dialect/FIRRTL/Transforms/ExtractInstances.cpp

```diff
@@ -0,0 +1,409 @@
+
+/// Emit the annotated source code for black boxes in a circuit.
+void ExtractInstancesPass::runOnOperation() {
+  CircuitOp circuitOp = getOperation();
+  anythingChanged = false;
+  anyFailures = false;
+  targetDir = ".";
+  annotatedModules.clear();
+  dutModules.clear();
+  extractionWorklist.clear();
+
+  // Determine the target directory from the annotations on the circuit
+  // operation.
+  for (auto anno : AnnotationSet(circuitOp)) {
+    // Handle target dir annotation.
+    if (anno.isClass(blackBoxTargetDirAnnoClass)) {
+      if (auto target = anno.getMember<StringAttr>("targetDir")) {
+        targetDir = target.getValue();
+        continue;
+      }
+      circuitOp->emitError(blackBoxTargetDirAnnoClass)
+          << " annotation missing \"targetDir\" attribute";
+      anyFailures = true;
+      continue;
+    }
+  }
+  if (anyFailures) {
+    signalPassFailure();
+    return;
+  }
+
+  // Walk the IR and gather all the annotations relevant for extraction that
+  // appear on instances and the instantiated modules.
+  instanceGraph = &getAnalysis<InstanceGraph>();
+  collectAnnos();
+  if (anyFailures) {
+    signalPassFailure();
+    return;
+  }
+
+  // Actually move instances upwards.
+  extractInstances();
+
+  // TODO: Group instances if so requested.
+
+  // If nothing has changed we can preseve the analysis.
+  LLVM_DEBUG(llvm::dbgs() << "\n");
+  if (!anythingChanged)
+    markAllAnalysesPreserved();
+  markAnalysesPreserved<InstanceGraph>();
+}
+
+static bool isAnnoInteresting(Annotation anno) {
+  return anno.isClass(extractBlackBoxAnnoClass);
+}
+
+/// Gather the modules and instances annotated to be moved by this pass. This
+/// populates the corresponding lists and maps of the pass.
+void ExtractInstancesPass::collectAnnos() {
+  CircuitOp circuit = getOperation();
+
+  // Gather the annotations on modules. These complement the later per-instance
+  // annotations.
+  circuit.walk([&](FModuleLike module) {
+    AnnotationSet::removeAnnotations(module, [&](Annotation anno) {
+      if (anno.isClass(dutAnnoClass)) {
+        LLVM_DEBUG(llvm::dbgs()
+                   << "Marking DUT `" << module.moduleName() << "`\n");
+        dutModules.insert(module);
+        return false; // other passes may rely on this anno; keep it
+      }
+      if (!isAnnoInteresting(anno))
+        return false;
+      LLVM_DEBUG(llvm::dbgs() << "Annotated module `" << module.moduleName()
+                              << "`:\n  " << anno.getDict() << "\n");
+      annotatedModules[module].push_back(anno);
+      return true;
+    });
+    return WalkResult::skip();
+  });
+
+  // Gather the annotations on instances to be extracted.
+  circuit.walk([&](InstanceOp inst) {
+    SmallVector<Annotation, 1> instAnnos;
+    Operation *module = instanceGraph->getReferencedModule(inst);
+
+    // Module-level annotations.
+    auto it = annotatedModules.find(module);
+    if (it != annotatedModules.end())
+      instAnnos.append(it->second);
```

Comment:
Something with count might be clearer:
```suggestion
    // Module-level annotations.
    if (annotatedModules.count(module))
      instAnnos.append(annotatedModules[module]);
```

Reply:
I usually do this because of performance paranoia, since using `count` and the index later have to do the lookup twice, wheras the iterator returned by `find` already points to the resolved location within the `DenseMap`, so accessing it for reading/writing is constant time.